runtime/tests/org.eclipse.core.tests.runtime/src/org/eclipse/core/tests/runtime/jobs/GithubBug_193.java

```diff
@@ -115,6 +147,99 @@ public String toString() {
 		}
 	}
 
+	@Test
+	public void testNestedSchedule() throws Exception {
+		int count = 10000;
+		for (int i = 0; i < count; i++) {
+			twoNestedJobs(i);
+		}
+	}
+
+	private void twoNestedJobs(int i) throws InterruptedException {
+		System.out.println("Starting " + i + " round");
+		int dummyJobsCount = i;
+		final int timeout = 10;
+		for (int j = 0; j < dummyJobsCount; j++) {
+			String name = "dummy_" + j;
+			Job job = Job.create(name, m -> {
+				// nada
+			});
+			if (i % 2 != 0) {
+				job.schedule(i % timeout);
+			} else {
+				job.schedule();
+			}
+			if (i % 2 == 0) {
+				job.join();
+			}
+		}
+
+		Job job1;
+		Job job2;
+		ReentrantLock lock = new ReentrantLock();
+		try {
+			lock.lock();
+
+			String jobName = "job_1_" + i;
+			job1 = new SillyJob(jobName, lock, timeout);
+			jobName = "job_2_" + i;
+			job2 = new SillyJob(jobName, lock, timeout);
+			Object syncObject = new Object();
+			job1.addJobChangeListener(new JobChangeAdapter() {
+				@Override
+				public void running(IJobChangeEvent event) {
+					job2.schedule();
+					synchronized (syncObject) {
+						syncObject.notify();
+					}
+				}
+			});
+
+			synchronized (syncObject) {
+				job1.schedule();
+//				Thread.sleep(10);
+				syncObject.wait();
+			}
+			assertEquals(Job.RUNNING, job1.getState());
+			assertEquals(Job.WAITING, job2.getState());
+			weirdCancel(timeout);
+		} finally {
+			lock.unlock();
+		}
+		Job.getJobManager().join(GithubBug_193.class, null);
+	}
+
+	private void weirdCancel(int timeout) {
+		try {
+			Thread.sleep(((int) (Math.random() * timeout / 2)));
+		} catch (InterruptedException e) {
+			// ignore
+		}
+		Job[] jobs = Job.getJobManager().find(null);
+		for (Job job : jobs) {
+			int state = job.getState();
```

Comment:
a decision on jobstate - which will change in parallel - is wrong

Reply:
We aren't talking about optimization of that test code here, do we?
The point is the hang, not on which criteria jobs are canceled. 

And the original code is not wrong, the code just optimizes execution of long running jobs that don't need to continue to do what they do (fetching data) if that data is not needed anymore.

Which is again not the point here and is not a root cause for the hang.